asserts/extkeypairmgr_test.go

```diff
@@ -0,0 +1,267 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts_test
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/asserts/assertstest"
+	"github.com/snapcore/snapd/testutil"
+)
+
+type extKeypairMgrSuite struct {
+	pgm *testutil.MockCmd
+
+	defaultPub *rsa.PublicKey
+	modelsPub  *rsa.PublicKey
+}
+
+var _ = Suite(&extKeypairMgrSuite{})
+
+func (s *extKeypairMgrSuite) SetUpSuite(c *C) {
+	tmpdir := c.MkDir()
+	k1, err := rsa.GenerateKey(rand.Reader, 4096)
+	c.Assert(err, IsNil)
+	k2, err := rsa.GenerateKey(rand.Reader, 4096)
+	c.Assert(err, IsNil)
+
+	derPub1, err := x509.MarshalPKIXPublicKey(&k1.PublicKey)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(tmpdir, "default.pub"), derPub1, 0644)
+	c.Assert(err, IsNil)
+	derPub2, err := x509.MarshalPKIXPublicKey(&k2.PublicKey)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(tmpdir, "models.pub"), derPub2, 0644)
+	c.Assert(err, IsNil)
+
+	err = ioutil.WriteFile(filepath.Join(tmpdir, "default.key"), x509.MarshalPKCS1PrivateKey(k1), 0600)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(tmpdir, "models.key"), x509.MarshalPKCS1PrivateKey(k2), 0600)
+	c.Assert(err, IsNil)
+
+	s.defaultPub = &k1.PublicKey
+	s.modelsPub = &k2.PublicKey
+
+	s.pgm = testutil.MockCommand(c, "keymgr", fmt.Sprintf(`
+keydir=%q
+case $1 in
+  features)
+    echo '{"signing":["RSA-PKCS"] , "public-keys":["DER"]}'
+    ;;
+  key-names)
+    echo '{"key-names": ["default", "models"]}'
+    ;;
+  get-public-key)
+    if [ "$5" = missing ]; then
+       echo not found
+       exit 1
+    fi
+    cat ${keydir}/"$5".pub
+    ;;
+  sign)
+    openssl rsautl -sign -pkcs -keyform DER -inkey ${keydir}/"$5".key
+    ;;
+  *)
+    exit 1
+    ;;
+esac
+`, tmpdir))
+}
+
+func (s *extKeypairMgrSuite) TearDownSuite(c *C) {
+	s.pgm.Restore()
+}
+
+func (s *extKeypairMgrSuite) TestFeaturesErrors(c *C) {
+	pgm := testutil.MockCommand(c, "keymgr", `
+if [ "$1" != "features" ]; then
+  exit 2
+fi
+if [ "${EXT_KEYMGR_FAIL}" = "exit-1" ]; then
+  exit 1
+fi
+echo "${EXT_KEYMGR_FAIL}"
+`)
+	defer pgm.Restore()
+	defer os.Unsetenv("EXT_KEYMGR_FAIL")
+
+	tests := []struct {
+		outcome string
+		err     string
+	}{
+		{"exit-1", `.*exit status 1.*`},
+		{`{"signing":["RSA-PKCS"]}`, `external keypair manager "keymgr" missing support for public key DER output format`},
+		{"{}", `external keypair manager \"keymgr\" missing support for RSA-PKCS signing`},
+		{"{", `cannot decode external keypair manager "keymgr" \[features\] output.*`},
+		{"", `cannot decode external keypair manager "keymgr" \[features\] output.*`},
+	}
+
+	for _, t := range tests {
+		os.Setenv("EXT_KEYMGR_FAIL", t.outcome)
```

Comment:
(nitpick^2) This environment var is not cleared AFAICT.

Reply:
cleared now in a defer